core/src/main/java/hudson/model/Executor.java

```diff
@@ -837,17 +838,37 @@ public void doStop( StaplerRequest req, StaplerResponse rsp ) throws IOException
     }
 
     /**
-     * Stops the current build.
+     * Stops the current build.<br>
+     * You can use {@link #doStopBuild(String)} instead to ensure what will be
+     * interrupted is actually what you want to interrupt.
      *
      * @since 1.489
+     * @see #doStopBuild(String)
      */
     @RequirePOST
     public HttpResponse doStop() {
+        return doStopBuild(null);
+    }
+
+    /**
+     * Stops the current build, if matching the specified external id
+     * (or no id is specified, or the current {@link Executable} is not a {@link Run}).
+     *
+     * @param runExtId
+     *      if not null, the externalizable id ({@link Run#getExternalizableId()})
+     *      of the build the user expects to interrupt
+     * @since TODO
+     */
+    @RequirePOST
+    public HttpResponse doStopBuild(@CheckForNull @QueryParameter(fixEmpty = true) String runExtId) {
```

Comment:
Does it have to be a public API? We could mark it as `@Restricted(NoExternalUse.class)` which is a good practice for stapler endpoints IMHO

Reply:
I think you're right, fixed in f01385e6a8.
FYI, to make sure this restriction was alright, I've searched through all existing uses of the original `Executor#doStop()` method by plugins in the jenkinsci organisation, in case some would have a reason to be changed to `Executor#doStopBuild(String)` instead. But the few call sites I've found are always similar to what is done in `AbstractBuild#doStop()`, with `Executor#doStop()` being called immediately on an executor which has just been retrieved as the current one for a specific build, so it would be redundant to check it's the "right" one twice.